modules/apps/static/portal-osgi-web/portal-osgi-web-servlet-context-helper-impl/src/main/java/com/liferay/portal/osgi/web/servlet/context/helper/internal/ServletContextHelperFactoryImpl.java

```diff
@@ -35,43 +37,52 @@
 public class ServletContextHelperFactoryImpl
 	implements ServletContextHelperFactory {
 
-	@Activate
-	protected void activate(
-			BundleContext bundleContext, Map<String, Object> properties)
-		throws Exception {
-
-		_saxParserFactory.setNamespaceAware(false);
-		_saxParserFactory.setValidating(false);
-		_saxParserFactory.setXIncludeAware(false);
-
-		try {
-			_saxParserFactory.setFeature(_FEATURES_DISALLOW_DOCTYPE_DECL, true);
-			_saxParserFactory.setFeature(
-				_FEATURES_EXTERNAL_GENERAL_ENTITIES, false);
-			_saxParserFactory.setFeature(
-				_FEATURES_EXTERNAL_PARAMETER_ENTITIES, false);
-			_saxParserFactory.setFeature(_FEATURES_LOAD_EXTERNAL_DTD, false);
-		}
-		catch (ParserConfigurationException | SAXNotRecognizedException |
-			   SAXNotSupportedException exception) {
-
-			ReflectionUtil.throwException(exception);
-		}
-
-		_executorService = _portalExecutorManager.getPortalExecutor(
-			ServletContextHelperFactoryImpl.class.getName());
-
-		_serviceRegistration = bundleContext.registerService(
-			ServletContextHelperRegistration.class.getName(),
-			new ServletContextHelperRegistrationServiceFactory(
-				_jspServletFactory, _saxParserFactory, properties,
-				_executorService),
-			null);
+	@Override
+	public void registerServletContextHelperRegistration() {
+		_serviceRegistrationDCLSingleton.getSingleton(
+			() -> {
+				_saxParserFactory.setNamespaceAware(false);
+				_saxParserFactory.setValidating(false);
+				_saxParserFactory.setXIncludeAware(false);
+
+				try {
+					_saxParserFactory.setFeature(
+						_FEATURES_DISALLOW_DOCTYPE_DECL, true);
+					_saxParserFactory.setFeature(
+						_FEATURES_EXTERNAL_GENERAL_ENTITIES, false);
+					_saxParserFactory.setFeature(
+						_FEATURES_EXTERNAL_PARAMETER_ENTITIES, false);
+					_saxParserFactory.setFeature(
+						_FEATURES_LOAD_EXTERNAL_DTD, false);
+				}
+				catch (ParserConfigurationException |
+					   SAXNotRecognizedException | SAXNotSupportedException
+						   exception) {
+
+					ReflectionUtil.throwException(exception);
+				}
+
+				_executorService = _portalExecutorManager.getPortalExecutor(
```

Comment:
@MrJohn1911 Having this in the singleton supplier creates a time window during which a NPE may happen if the component gets deactivated. I still prefer to leave this one in an `activate` method.